File: python3/leetcodepy/n_queens.py
"""
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return all distinct solutions to the n-queens puzzle.

Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a queen and an empty space, respectively.



Example 1:

https://assets.leetcode.com/uploads/2020/11/13/queens.jpg
Input: n = 4
Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
Explanation: There exist two distinct solutions to the 4-queens puzzle as shown above

Example 2:

Input: n = 1
Output: [["Q"]]


Constraints:

1 <= n <= 9
"""
from typing import List


class Solution1:
    def solveNQueens(self, n: int) -> List[List[str]]:
        col_used = [False] * n
        diagonal_used, anti_diagonal_used = [False] * (2 * n - 1), [False] * (2 * n - 1)
        result = []
        path = [['.'] * n for _ in range(n)]
        self.solve_n_queens(n, 0, col_used, diagonal_used, anti_diagonal_used, result, path)
        return result

    def solve_n_queens(self, n: int, i: int,
                       col_used: List[bool], diagonal_used: List[bool], anti_diagonal_used: List[bool],
                       result: List[List[str]], path: List[List[str]]):
        if i == n:
            result.append([''.join(row) for row in path])
            return
        for j in range(n):
            d, a = n - 1 + i - j, i + j
            if col_used[j] or diagonal_used[d] or anti_diagonal_used[a]:
                continue
            col_used[j] = diagonal_used[d] = anti_diagonal_used[a] = True
            path[i][j] = 'Q'
            self.solve_n_queens(n, i + 1, col_used, diagonal_used, anti_diagonal_used, result, path)
            path[i][j] = '.'
            col_used[j] = diagonal_used[d] = anti_diagonal_used[a] = False

File: java/src/main/java/io/binac/leetcode/NQueens.java
package io.binac.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * <p>The <strong>n-queens</strong> puzzle is the problem of placing <code>n</code> queens on an <code>n x n</code> chessboard such that no two queens attack each other.</p>
 *
 * <p>Given an integer <code>n</code>, return <em>all distinct solutions to the <strong>n-queens puzzle</strong></em>.</p>
 *
 * <p>Each solution contains a distinct board configuration of the n-queens' placement, where <code>'Q'</code> and <code>'.'</code> both indicate a queen and an empty space, respectively.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2020/11/13/queens.jpg" style="width: 600px; height: 268px;">
 * <pre><strong>Input:</strong> n = 4
 * <strong>Output:</strong> [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
 * <strong>Explanation:</strong> There exist two distinct solutions to the 4-queens puzzle as shown above
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> n = 1
 * <strong>Output:</strong> [["Q"]]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= n &lt;= 9</code></li>
 * </ul>
 */
public class NQueens {
    public static class Solution1 {
        private void solveNQueens(char[][] board, int n, int i, List<List<String>> result,
                                  boolean[] col, boolean[] mainDiagonal, boolean[] antiDiagonal) {
            if (i == n) {
                List<String> solution = new ArrayList<>(n);
                for (char[] chars : board)
                    solution.add(new String(chars));
                result.add(solution);
                return;
            }
            for (int j = 0, d, a; j < n; ++j) {
                d = n - 1 + i - j;
                a = i + j;
                if (col[j] || mainDiagonal[d] || antiDiagonal[a]) continue;
                board[i][j] = 'Q';
                col[j] = true;
                mainDiagonal[d] = true;
                antiDiagonal[a] = true;
                solveNQueens(board, n, i + 1, result, col, mainDiagonal, antiDiagonal);
                col[j] = false;
                mainDiagonal[d] = false;
                antiDiagonal[a] = false;
                board[i][j] = '.';
            }
        }

        public List<List<String>> solveNQueens(int n) {
            List<List<String>> result = new ArrayList<>();
            char[][] board = new char[n][n];
            for (int i = 0; i < n; ++i) Arrays.fill(board[i], '.');
            solveNQueens(board, n, 0, result, new boolean[n], new boolean[2 * n - 1], new boolean[2 * n - 1]);
            return result;
        }
    }
}
